draft-barnes-mls-protocol.md

```diff
@@ -1435,44 +1434,204 @@ all arrive at the following state:
-group key exchange among participants.  In order to protect messages sent among
-those participants, an application will need to specify how messages are protected.
+The traffic keying material is generated from an input traffic secret value using:
 
-For every epoch, the root key of the ratcheting tree can be used to
-derive key material for symmetric operations such as encryption/AEAD and MAC;
-AEAD or MAC MUST be used to ensure that the message originated from a member
-of the group.
+~~~~
+[sender]_write_key =
+  HKDF-Expand-Label(Application_Secret,"mls app key", "", key_length)
 
-In addition, asymmetric signatures SHOULD be used to authenticate the sender
-of a message.
+[sender]_write_iv  =
+  HKDF-Expand-Label(Application_Secret,"mls app iv", "", iv_length)
+~~~~
 
-In combination with server-side enforced ordering, data from previous messages
-is used (as a salt when hashing) to:
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitely inherit this ownership.
 
- * add freshness to derived symmetric keys
- * cryptographically bind the transcript of all previous messages with the current
-   group shared secret
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
 
-Possible candidates for that are:
+Updating the Application secret and deriving the associated AEAD key and IV can
+be summerized as the following Application key schedule:
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+## Application Key Schedule {#key-schedule-application}
+
+Each participant Application secret chain looks as follows after the initial
+derivation:
+
+~~~~~
+           Application Secret [n-1]
+                     |
+                     +--> HKDF-Expand-Label(.,"mls app key", "", key_length)
+                     |    = Application AEAD Key [n-1]
+                     |
+                     +--> HKDF-Expand-Label(.,"mls app iv", "", iv_length)
+                     |    = Application AEAD IV [n-1]
+                     |
+                     V
+           HKDF-Expand-Label(.,"mls app upd", App_Context, Hash.length)
+                     |
+                     V
+           Application Secret [n]
+                     |
+                     +--> HKDF-Expand-Label(.,"mls app key", "", key_length)
+                     |    = Application AEAD Key [n]
+                     |
+                     +--> HKDF-Expand-Label(.,"mls app iv", "", iv_length)
+                          = Application AEAD IV [n]
+~~~~~
+
+[[ OPEN ISSUE: At the moment there is no contributivity of Application secrets
+chained from the initial one to the next generation of Epoch secret.]]
```

Comment:
My intuition is that we don't need to do it because the properties we will get on the group state can't be influenced by the operations that use the application secret or its derivations...